contracts/PriceOracle/PriceOracle.sol

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+import {Unauthorized} from "../common/Errors.sol";
+import {System} from "../common/System/System.sol";
+import {HydraChainConnector} from "../HydraChain/HydraChainConnector.sol";
+import {APRCalculatorConnector} from "../APRCalculator/APRCalculatorConnector.sol";
+import {IPriceOracle, PriceForValidator} from "./IPriceOracle.sol";
+
+/**
+ * @title PriceOracle
+ * @dev This contract will be responsible for the price updates.
+ * Active validators will be able to vote and agree on the price.
+ */
+contract PriceOracle is IPriceOracle, System, Initializable, HydraChainConnector, APRCalculatorConnector {
+    mapping(uint256 => PriceForValidator[]) public priceVotesForDay;
+    mapping(address => uint256) public validatorLastVotedDay;
+    mapping(uint256 => uint256) public pricePerDay;
+
+    // _______________ Initializer _______________
+
+    function initialize(address _hydraChainAddr, address _aprCalculatorAddr) external initializer onlySystemCall {
+        __HydraChainConnector_init(_hydraChainAddr);
+        __APRCalculatorConnector_init(_aprCalculatorAddr);
+    }
+
+    // _______________ Modifers _______________
+
+    modifier onlyActiveValidator() {
+        if (hydraChainContract.isValidatorActive(msg.sender) == false) {
+            revert Unauthorized("INACTIVE_STAKER");
+        }
+        _;
+    }
+
+    // _______________ External functions _______________
+
+    /**
+     * @inheritdoc IPriceOracle
+     */
+    function vote(uint256 _price) external {
+        uint256 day = getTodayIfVoteAvailable(_price);
+        validatorLastVotedDay[msg.sender] = day;
+
+        PriceForValidator memory priceForValidator = PriceForValidator({price: _price, validator: msg.sender});
+        priceVotesForDay[day].push(priceForValidator);
+
+        emit PriceVoted(_price, msg.sender, day);
+
+        uint256 price = _checkPriceUpdateAvailability(day);
+
+        if (price != 0) {
+            _updatePrice(price, day);
+        }
+    }
+
+    // _______________ Public functions _______________
+
+    function getTodayIfVoteAvailable(uint256 _price) public view onlyActiveValidator returns (uint256 day) {
+        day = _getCurrentDay();
+        if (_price == 0) {
+            revert InvalidPrice();
+        }
+
+        if (pricePerDay[day] != 0) {
+            revert PriceAlreadySet();
+        }
+
+        if (validatorLastVotedDay[msg.sender] == day) {
+            revert AlreadyVoted();
+        }
+    }
+
+    // _______________ Internal functions _______________
+
+    /**
+     * @notice Check if the price update is available for the given day
+     * @param _day Day to check
+     * @return uint256 Price if available, 0 otherwise
+     */
+    function _checkPriceUpdateAvailability(uint256 _day) internal view returns (uint256) {
```

Comment:
You must keep a record of the price group with the highest voting power and for a new pending one (if there is more than one). Update the best group if the pending reaches more voting power than the previous best. That way in the end, the best will be the winner price.